File: 13-advanced/yegge/ruby/yegge.rb
# def reverse(str)
#   str.reverse
# end

# Iterative
# def reverse(str)
#   output = ''
#   str.chars.each { |c| output = c + output } # There is an evil solution using reduce()
#   output
# end

# Recursive
def reverse(str)
  if str.empty?
    str # Base case
  else
    str.chars.last + reverse( str[0...str.length - 1] )
  end
end

def twelve_times_table
  (0..12).each do |m|
    puts "#{ m } x 12 = #{ m * 12 }" # sprintf("%d x 12 = %d", m, m * 12) 
  end
end

require 'pry'
binding.pry

File: 13-advanced/yegge/js/yegge.js
// Write a function to reverse a string.

const lazyReverse = function (s) {
  // This is cheating?
  return s.split('').reverse().join('');
}

console.log('SoMeStRiNg', lazyReverse('SoMeStRiNg'));

// Iterative reverse
// const reverse = function (s) {
//   let output = '';
//   for (let i = s.length - 1; i >= 0; i--) {
//     output += s[i]; // Add character to the end of output
//   }
//   return output;
// }

const reverse = function (s) {
  let output = '';
  s.split('').forEach((c) => {
    output = c + output; // Add character to the beginning of output.
  });
  return output;
}

console.log('SoMeStRiNg', reverse('SoMeStRiNg'));

// Recursive reverse
const recursiveReverse = function (s) {
  if (s === '') {
    return ''; // Base case
  } else {
    return s[ s.length - 1] + recursiveReverse( s.slice(0, s.length - 1 ) );
  }
}

console.log('SoMeStRiNg', recursiveReverse('SoMeStRiNg'));
